weldx/asdf/util.py

```diff
@@ -288,3 +288,69 @@ def view_tree(file: types_path_and_file_like, path: Tuple = None, **kwargs):
 def asdf_json_repr(file: Union[str, Path, BytesIO], path: Tuple = None, **kwargs):
     """See `view_tree` function."""
     return view_tree(file, path, **kwargs)
+
+
+def dataclass_serialization_class(
+    class_type: Type,
+    class_name: str,
+    version: str,
+    to_tree_mod: Callable = None,
+    from_tree_mod: Callable = None,
+) -> Type:
+    """Generate a asdf serialization class for a python dataclass.
+
+    Parameters
+    ----------
+    class_type :
+        The type of the dataclass
+    class_name :
+        The value that should ba stored as the classes name property
+    version :
+        The version number
+    to_tree_mod :
+        A method that applies additional modifications to the tree during the `to_tree`
+        function call
+    from_tree_mod :
+        A method that applies additional modifications to the tree during the
+        `from_tree` function call
+
+    Returns
+    -------
+    Type :
+        A new asdf serialization class.
+
+    """
+    from copy import deepcopy
+
+    import numpy as np
+
+    from weldx.asdf.types import WeldxType
+
+    def _noop(tree):
+        return tree
+
+    if to_tree_mod is None:
+        to_tree_mod = _noop
+    if from_tree_mod is None:
+        from_tree_mod = _noop
+
+    class _SerializationClass(WeldxType):
+        name = class_name
+        version = "1.0.0"
+        types = [class_type]
+        requires = ["weldx"]
+        handle_dynamic_subclasses = True
+
+        @classmethod
+        def to_tree(cls, node, ctx):
+            return to_tree_mod(deepcopy(node.__dict__))
```

Comment:
I don't think `deepcopy` should be the default, would that not break links/refs in asdf files?

Reply:
The problem here is that if you modify the dict, you also modify the dataclass. Due to `to_tree` being called twice per write, this causes errors whenever you need to modify the dict before passing it to asdf. However, we can probably remove it as default and just add the copy whenever we provide a modification function.